src/transform.js

```diff
@@ -0,0 +1,40 @@
+import { properties as childProperties } from "./children.js";
+import { matches } from "./util.js";
+
+
+/**
+ * Recursively execute a callback on this node and all its children.
+ * If the callback returns a non-undefined value, it will overwrite the node.
+ * 
+ * @param {object | object[]} node AST node or array of nodes
+ * @param {function(object, string, object?)} transformCallback
+ * @param {object} [o]
+ * @param {string | string[] | function} [o.only] Only walk nodes of this type
+ * @param {string | string[] | function} [o.except] Ignore walking nodes of these types
+ */
+export default function transform (node, transformCallback, o) {
+	_transform(node, transformCallback, o);
+}
+
+function _transform (node, callback, o = {}, property, parent) {
+	if (Array.isArray(node)) {
+		return node.forEach(n => _transform(n, callback, o, property, parent));
+	}
+
+	const ignore = o.except && matches(node, o.except);
+	const explore = !ignore && matches(node, o.only);
+	
+	if (explore) {
+		const transformedNode = callback(node, property, parent);
+		if (transformedNode !== undefined) {
+			// delete all keys in node:
```

Comment:
As it turns out, we actually have to do it using the `Object.assign` approach. The other approach where we reassign the local reference to `node` only works when aliasing comes into play with a tree of depth > 1. So for a simple rewrite of an expression like `foo`, the other approach will not succeed (I added a test case to check for this).

Reply:
Could you expand on that? I think modifying the objects is suboptimal, so I'd rather avoid it if possible.

Reply:
So for example if we have a single node tree, `{type: "Identifier", name: "foo"}`, in the other approach, where we do `node = transformedNode !== undefined ? transformedNode : node`, reassigning the local reference won't actually mutate the node since there are no child properties that are explored and mutated. I couldn't think of another way to do this correctly without modifying the keys of `node` in some way.

Reply:
Ah, I see. So the issue is when the whole AST is mutated, right? Everything else will have a parent-child relationship.

Reply:
Yes

Reply:
I think it's fine to return a new node in that case, and just make sure to document that caveat.

Reply:
The issue is that returning a new node runs into the issue that lead to implementing the Object.assign issue in the first place, where trees of height 1 don't get transformed. Alternatively, another approach would be implement `map`, and then implement transform like this:

```js
function transform (node, callback, o) {
	const ret = map(node, callback, o);
        Object.assign(node, ret);
}
```

Given that `map` has a much cleaner solution, this might be a potential direction we could go. What do you think?

Reply:
Like I said, I think gutting an object and replacing all its properties with properties from another object is way messier than documenting a caveat that if you’re transforming the root node, you need to take care of that. It's also quite unexpected, that's not how JS code usually works. Whereas just having the restriction that if you're transforming the top-level node, that will not actually be done by reference, seems quite expected (to me at least), and pretty easy to work around by the user if it's not the desired behavior.

Reply:
Oh ok I see what you mean, I will revert back to the old version

Reply:
Just reverted back

Reply:
Thanks! 

Basically, this is something you'd do as a last resort (and I think it's a code smell that we had to do it in a part of Mavo). When it's objects you don't own, it's hard (in some cases even impossible) to do it 100% right in all edge cases, and even when it's objects you do own, it's unusual, and thus a bit unexpected and could break assumptions in calling code. It's a nice hack and it's good to know that it's possible, but like all hacks, it's good to explore alternatives first 😊

Reply:
Makes sense, and good to know for the future. Any other feedback for this PR?